### PR TITLE
feat: parallel update

### PR DESCRIPTION
**Pull request by rphmeier**
just a skeleton for now

### CODE REVIEW

**Comment by rphmeier** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/170?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#170** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/170?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#166** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#165** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#161** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/161?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/lib.rs [hunk 1]
```diff
@@ -165,12 +173,9 @@ impl Nomt {
         let page_cache = PageCache::new(store.clone(), &o);
         let root = store.load_root()?;
         Ok(Self {
+            commit_pool: CommitPool::new(o.fetch_concurrency),
```
Comment (pepyakin) [inline nomt/src/lib.rs:176]: it came to my attention that `o.traversal_concurrency` is not really used anymore

Comment (rphmeier) [reply nomt/src/lib.rs:176]: yes, I noticed this as well. we lost it somewhere, but it was before this stack.

**Review by pepyakin — APPROVED** [review]

File: nomt/src/commit/mod.rs [hunk 2]
```diff
@@ -0,0 +1,429 @@
+//! Parallel, pipelined merkle trie commits.
+//!
+//! This splits the work of warming-up and performing the trie update across worker threads.
+
+#![allow(unused)] // TODO
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply nomt/src/commit/mod.rs:5]: d'oh.

thanks

File: nomt/src/commit/worker.rs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    shared: Arc<CommitShared>,
+    write_pass: WritePass<PageRegion>,
+    page_walker: PageWalker<H>,
+    page_cache: PageCache,
+    range_start: usize,
+    range_end: usize,
+    cur_index: usize,
+
+    // pending trie position to be compacted upwards.
+    pending: Option<TriePosition>,
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply nomt/src/commit/worker.rs:190]: vestigial from before I realized I needed to create the pagewalker.

File: nomt/src/commit/worker.rs [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            let mut batch_size = 0;
+            let mut has_writes = false;
+
+            // find batch size.
+            for (k, v) in &self.shared.read_write[start_index..] {
+                if !seek_result.position.subtrie_contains(&k) {
+                    break;
+                }
+                batch_size += 1;
+                has_writes = v.is_write();
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply nomt/src/commit/worker.rs:268]: yes, wow. nice one.

[the remark this answers is not in the document]
Comment (pepyakin) [reply nomt/src/commit/worker.rs]: wonder why tests didnt catch this...

**Pull request merged**